File: Sources/IrohaCore/UserDictionaryStore.swift
import Foundation

/// ユーザ辞書の永続化（JSONファイル）。
///
/// irohaのユーザ辞書はこのファイルが唯一の情報源で、macOSのユーザ辞書からは
/// `syncFromSystem(_:)`で取り込むだけ（macOS側には一切書き込まない）。
/// IMEと設定ウィンドウは同一プロセスなので、`shared`をプロセス内で共有すれば足りる。
public final class UserDictionaryStore: @unchecked Sendable {

    /// 内容が変わったときに通知する（設定ウィンドウの一覧更新用）
    public static let didChangeNotification = Notification.Name("iroha.userDictionaryDidChange")

    public static let defaultURL = URL(
        fileURLWithPath: NSHomeDirectory()
            + "/Library/Application Support/iroha/user-dictionary.json")

    public static let shared = UserDictionaryStore()

    /// 取り込み結果のサマリ（設定UIでの報告用）
    public struct SyncResult: Sendable, Equatable {
        public var added: Int
        public var removed: Int
        public var unchanged: Int
        public var skipped: Int   // 読みがひらがなでない等で対象外にしたもの
    }

    private let url: URL
    private let lock = NSLock()
    private var cached: UserDictionary

    public init(url: URL = UserDictionaryStore.defaultURL) {
        self.url = url
        self.cached = Self.load(from: url)
    }

    /// 変換時に参照するスナップショット
    public var current: UserDictionary {
        lock.lock()
        defer { lock.unlock() }
        return cached
    }

    public var entries: [UserDictionaryEntry] { current.entries }

    /// 一覧をまるごと置き換えて保存する（設定UIの編集結果の反映）
    @discardableResult
    public func replaceAll(_ entries: [UserDictionaryEntry]) -> UserDictionary {
        store(entries)
    }

    @discardableResult
    public func add(reading: String, word: String) -> UserDictionary {
        var entries = self.entries
        entries.append(UserDictionaryEntry(reading: reading, word: word, source: .manual))
        return store(entries)
    }

    @discardableResult
    public func remove(ids: Set<UUID>) -> UserDictionary {
        store(entries.filter { !ids.contains($0.id) })
    }

    /// macOSのユーザ辞書の内容に合わせる。
    ///
    /// - `.system`（取り込み由来）のエントリだけを追加・削除の対象にし、
    ///   irohaで追加・編集した`.manual`のエントリには触らない
    /// - macOS側から消えたエントリはこちらからも消す（ミラーリング）
    @discardableResult
    public func syncFromSystem(_ systemEntries: [(reading: String, word: String)]) -> SyncResult {
        var valid: [(reading: String, word: String)] = []
        var skipped = 0
        for entry in systemEntries {
            let reading = UserDictionary.normalizedReading(entry.reading)
            let word = entry.word.trimmingCharacters(in: .whitespacesAndNewlines)
            guard UserDictionary.isImportableReading(reading), !word.isEmpty else {
                skipped += 1
                continue
            }
            valid.append((reading, word))
        }

        lock.lock()
        let existing = cached.entries
        lock.unlock()

        // 手動エントリと同じ内容が既にあるものは追加しない（重複表示を避ける）
        let manualPairs = Set(existing.filter { $0.source == .manual }.map { Pair($0) })
        var wanted: [Pair: Void] = [:]
        var ordered: [Pair] = []
        for entry in valid {
            let pair = Pair(reading: entry.reading, word: entry.word)
            guard !manualPairs.contains(pair), wanted[pair] == nil else { continue }
            wanted[pair] = ()
            ordered.append(pair)
        }

        var result = SyncResult(added: 0, removed: 0, unchanged: 0, skipped: skipped)
        var updated: [UserDictionaryEntry] = []
        var kept: Set<Pair> = []
        for entry in existing {
            guard entry.source == .system else {
                updated.append(entry)
                continue
            }
            let pair = Pair(entry)
            if wanted[pair] != nil, !kept.contains(pair) {
                updated.append(entry)
                kept.insert(pair)
                result.unchanged += 1
            } else {
                result.removed += 1
            }
        }
        for pair in ordered where !kept.contains(pair) {
            updated.append(
                UserDictionaryEntry(reading: pair.reading, word: pair.word, source: .system))
            result.added += 1
        }

        if result.added > 0 || result.removed > 0 {
            store(updated)
        }
        return result
    }

    // MARK: - 内部

    /// (読み, 単語)の同一性判定用
    private struct Pair: Hashable {
        var reading: String
        var word: String
        init(reading: String, word: String) {
            self.reading = reading
            self.word = word
        }
        init(_ entry: UserDictionaryEntry) {
            self.init(reading: entry.reading, word: entry.word)
        }
    }

    @discardableResult
    private func store(_ entries: [UserDictionaryEntry]) -> UserDictionary {
        // 読みが空・単語が空のものは落とし、読み順に並べる（同じ読みの候補は登録順を保つ）
        let cleaned = entries
            .map {
                UserDictionaryEntry(
                    id: $0.id,
                    reading: UserDictionary.normalizedReading($0.reading),
                    word: $0.word.trimmingCharacters(in: .whitespacesAndNewlines),
                    source: $0.source)
            }
            .filter { !$0.reading.isEmpty && !$0.word.isEmpty }
        let sorted = cleaned.enumerated()
            .sorted { ($0.element.reading, $0.offset) < ($1.element.reading, $1.offset) }
            .map(\.element)

        let dictionary = UserDictionary(entries: sorted)
        lock.lock()
        cached = dictionary
        lock.unlock()
        save(sorted)
        // 受け手はUI（設定ウィンドウ）なのでメインスレッドに揃える
        // （起動時の自動同期はバックグラウンドから呼ばれる）
        DispatchQueue.main.async {
            NotificationCenter.default.post(name: Self.didChangeNotification, object: nil)
        }
        return dictionary
    }

    private struct FileContents: Codable {
        var version: Int
        var entries: [UserDictionaryEntry]
    }

    private static func load(from url: URL) -> UserDictionary {
        guard let data = try? Data(contentsOf: url),
              let contents = try? JSONDecoder().decode(FileContents.self, from: data)
        else { return .empty }
        return UserDictionary(entries: contents.entries)
    }

    private func save(_ entries: [UserDictionaryEntry]) {
        do {
            try FileManager.default.createDirectory(
                at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
            let data = try encoder.encode(FileContents(version: 1, entries: entries))
            try data.write(to: url, options: .atomic)
        } catch {
            NSLog("iroha: ユーザ辞書の保存に失敗: \(error)")
        }
    }
}

File: windows/iroha-core/src/user_dictionary_store.cpp
#include "iroha/user_dictionary_store.h"

#include <algorithm>
#include <fstream>
#include <set>
#include <sstream>

#include "iroha/json.h"
#include "iroha/unicode.h"

namespace iroha {

namespace {

std::vector<UserDictionaryEntry> Load(const std::filesystem::path& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) return {};
    std::stringstream buffer;
    buffer << file.rdbuf();
    json::Value root;
    if (!json::Parse(buffer.str(), &root)) return {};
    const json::Value* entriesValue = root.Find("entries");
    if (!entriesValue || entriesValue->type != json::Value::Type::Array) return {};

    std::vector<UserDictionaryEntry> entries;
    for (const json::Value& item : entriesValue->array) {
        const json::Value* reading = item.Find("reading");
        const json::Value* word = item.Find("word");
        if (!reading || !word) continue;
        UserDictionaryEntry entry;
        entry.reading = Utf8ToUtf32(reading->AsString());
        entry.word = Utf8ToUtf32(word->AsString());
        if (const json::Value* id = item.Find("id")) entry.id = id->AsString();
        if (entry.id.empty()) entry.id = GenerateUuid();
        if (const json::Value* source = item.Find("source")) {
            entry.source = source->AsString() == "system"
                               ? UserDictionaryEntry::Source::System
                               : UserDictionaryEntry::Source::Manual;
        }
        entries.push_back(std::move(entry));
    }
    return entries;
}

using Pair = std::pair<std::u32string, std::u32string>; // (読み, 単語)の同一性判定用

} // namespace

UserDictionaryStore::UserDictionaryStore(std::filesystem::path path)
    : path_(std::move(path)), cached_(Load(path_)) {}

void UserDictionaryStore::ReplaceAll(std::vector<UserDictionaryEntry> entries) {
    Store(std::move(entries));
}

void UserDictionaryStore::Add(const std::u32string& reading,
                              const std::u32string& word) {
    std::vector<UserDictionaryEntry> entries = cached_.Entries();
    entries.push_back(
        MakeUserDictionaryEntry(reading, word, UserDictionaryEntry::Source::Manual));
    Store(std::move(entries));
}

UserDictionaryStore::SyncResult UserDictionaryStore::SyncFromSystem(
    const std::vector<std::pair<std::u32string, std::u32string>>& systemEntries) {
    std::vector<Pair> valid;
    int skipped = 0;
    for (const auto& [rawReading, rawWord] : systemEntries) {
        const std::u32string reading = UserDictionary::NormalizedReading(rawReading);
        const std::u32string word = TrimWhitespace(rawWord);
        if (!UserDictionary::IsImportableReading(reading) || word.empty()) {
            ++skipped;
            continue;
        }
        valid.push_back({reading, word});
    }

    const std::vector<UserDictionaryEntry>& existing = cached_.Entries();

    // 手動エントリと同じ内容が既にあるものは追加しない（重複表示を避ける）
    std::set<Pair> manualPairs;
    for (const UserDictionaryEntry& entry : existing) {
        if (entry.source == UserDictionaryEntry::Source::Manual) {
            manualPairs.insert({entry.reading, entry.word});
        }
    }
    std::set<Pair> wanted;
    std::vector<Pair> ordered;
    for (const Pair& pair : valid) {
        if (manualPairs.count(pair) || wanted.count(pair)) continue;
        wanted.insert(pair);
        ordered.push_back(pair);
    }

    SyncResult result;
    result.skipped = skipped;
    std::vector<UserDictionaryEntry> updated;
    std::set<Pair> kept;
    for (const UserDictionaryEntry& entry : existing) {
        if (entry.source != UserDictionaryEntry::Source::System) {
            updated.push_back(entry);
            continue;
        }
        const Pair pair{entry.reading, entry.word};
        if (wanted.count(pair) && !kept.count(pair)) {
            updated.push_back(entry);
            kept.insert(pair);
            ++result.unchanged;
        } else {
            ++result.removed;
        }
    }
    for (const Pair& pair : ordered) {
        if (kept.count(pair)) continue;
        updated.push_back(MakeUserDictionaryEntry(pair.first, pair.second,
                                                  UserDictionaryEntry::Source::System));
        ++result.added;
    }

    if (result.added > 0 || result.removed > 0) {
        Store(std::move(updated));
    }
    return result;
}

void UserDictionaryStore::Store(std::vector<UserDictionaryEntry> entries) {
    // 読みが空・単語が空のものは落とし、読み順に並べる（同じ読みの候補は登録順を保つ）
    std::vector<UserDictionaryEntry> cleaned;
    for (UserDictionaryEntry& entry : entries) {
        entry.reading = UserDictionary::NormalizedReading(entry.reading);
        entry.word = TrimWhitespace(entry.word);
        if (entry.reading.empty() || entry.word.empty()) continue;
        cleaned.push_back(std::move(entry));
    }
    std::stable_sort(cleaned.begin(), cleaned.end(),
                     [](const UserDictionaryEntry& a, const UserDictionaryEntry& b) {
                         return a.reading < b.reading;
                     });
    Save(cleaned);
    cached_ = UserDictionary(std::move(cleaned));
}

void UserDictionaryStore::Save(const std::vector<UserDictionaryEntry>& entries) const {
    // macOS版のJSONEncoder（.prettyPrinted, .sortedKeys, .withoutEscapingSlashes）と同形式
    std::string out = "{\n  \"entries\" : [\n";
    for (size_t i = 0; i < entries.size(); ++i) {
        const UserDictionaryEntry& entry = entries[i];
        out += "    {\n";
        out += "      \"id\" : " + json::QuoteString(entry.id) + ",\n";
        out += "      \"reading\" : " + json::QuoteString(Utf32ToUtf8(entry.reading)) +
               ",\n";
        out += "      \"source\" : " +
               json::QuoteString(entry.source == UserDictionaryEntry::Source::System
                                     ? "system"
                                     : "manual") +
               ",\n";
        out += "      \"word\" : " + json::QuoteString(Utf32ToUtf8(entry.word)) + "\n";
        out += i + 1 < entries.size() ? "    },\n" : "    }\n";
    }
    out += "  ],\n  \"version\" : 1\n}";

    std::error_code ec;
    std::filesystem::create_directories(path_.parent_path(), ec);
    const std::filesystem::path tmp = path_.wstring() + L".tmp";
    {
        std::ofstream file(tmp, std::ios::binary | std::ios::trunc);
        if (!file) return;
        file.write(out.data(), static_cast<std::streamsize>(out.size()));
    }
    std::filesystem::rename(tmp, path_, ec);
}

} // namespace iroha
